File: lib/norikra/udf/woothee.rb
require "norikra/udf/woothee/version"

require 'java'
require 'norikra/udf'

module Norikra
  module UDF
    # class style
    class WootheeParseAgent < Norikra::UDF::SingleRow
      def self.init
        require 'woothee.jar'
        require 'norikra-udf-woothee.jar'
      end
      def function_name
        "parseAgent"
      end
      def class_name
        "is.tagomor.norikra.udf.Woothee"
      end
      def method_name # default: same as function_name
        "parseAgent"
      end
      # "definition" also available
    end

    # udf bundle module style
    module WootheeUDF
      def self.init
        require 'woothee.jar'
        require 'norikra-udf-woothee.jar'
      end

      def self.plugins
        [IsPC, IsSmartPhone, IsMobilePhone, IsAppliance, IsMisc, IsCrawler, IsUnknown, IsIn]
      end

      class IsPC < Norikra::UDF::SingleRow
        def definition
          ["isPC", "is.tagomor.norikra.udf.Woothee", "isPC"]
        end
      end
      class IsSmartPhone < Norikra::UDF::SingleRow
        def definition
          ["isSmartPhone", "is.tagomor.norikra.udf.Woothee", "isSmartPhone"]
        end
      end
      class IsMobilePhone < Norikra::UDF::SingleRow
        def definition
          ["isMobilePhone", "is.tagomor.norikra.udf.Woothee", "isMobilePhone"]
        end
      end
      class IsAppliance < Norikra::UDF::SingleRow
        def definition
          ["isAppliance", "is.tagomor.norikra.udf.Woothee", "isAppliance"]
        end
      end
      class IsMisc < Norikra::UDF::SingleRow
        def definition
          ["isMisc", "is.tagomor.norikra.udf.Woothee", "isMisc"]
        end
      end
      class IsCrawler < Norikra::UDF::SingleRow
        def definition
          ["isCrawler", "is.tagomor.norikra.udf.Woothee", "isCrawler"]
        end
      end
      class IsUnknown < Norikra::UDF::SingleRow
        def definition
          ["isUnknown", "is.tagomor.norikra.udf.Woothee", "isUnknown"]
        end
      end
      class IsIn < Norikra::UDF::SingleRow
        def definition
          ["isIn", "is.tagomor.norikra.udf.Woothee", "isIn"]
        end
      end
    end
  end
end

File: java/is/tagomor/norikra/udf/Woothee.java
package is.tagomor.norikra.udf;

import is.tagomor.woothee.Classifier;
import is.tagomor.woothee.DataSet;

import java.util.Map;
import java.util.List;

public final class Woothee
{
  public static Map<String,String> parseAgent(final String agent) {
    return Classifier.parse(agent);
  }

  public static Boolean isPC(final Map<String,String> m) {
    return m.get(DataSet.ATTRIBUTE_CATEGORY).equals(DataSet.DATASET_CATEGORY_PC);
  }

  public static Boolean isSmartPhone(final Map<String,String> m) {
    return m.get(DataSet.ATTRIBUTE_CATEGORY).equals(DataSet.DATASET_CATEGORY_SMARTPHONE);
  }

  public static Boolean isMobilePhone(final Map<String,String> m) {
    return m.get(DataSet.ATTRIBUTE_CATEGORY).equals(DataSet.DATASET_CATEGORY_MOBILEPHONE);
  }

  public static Boolean isAppliance(final Map<String,String> m) {
    return m.get(DataSet.ATTRIBUTE_CATEGORY).equals(DataSet.DATASET_CATEGORY_APPLIANCE);
  }

  public static Boolean isMisc(final Map<String,String> m) {
    return m.get(DataSet.ATTRIBUTE_CATEGORY).equals(DataSet.DATASET_CATEGORY_MISC);
  }

  public static Boolean isCrawler(final Map<String,String> m) {
    return m.get(DataSet.ATTRIBUTE_CATEGORY).equals(DataSet.DATASET_CATEGORY_CRAWLER);
  }

  public static Boolean isUnknown(final Map<String,String> m) {
    return m.get(DataSet.ATTRIBUTE_CATEGORY).equals(DataSet.VALUE_UNKNOWN);
  }

  public static Boolean isIn(final Map<String,String> m, final List<String> categories) {
    String targetCategory = m.get(DataSet.ATTRIBUTE_CATEGORY);
    for (String c : categories) {
      if (targetCategory.equals(c))
        return Boolean.TRUE;
    }
    return Boolean.FALSE;
  }
}
